{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3cfe930e",
   "metadata": {},
   "source": [
    "## Running multiple SigMApy analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e1f899ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input directories:\n",
      "['/home/ferperez/SigMA_tests/Input_datasets/BRCA/mc3', '/home/ferperez/SigMA_tests/Input_datasets/BRCA/Foundation_One', '/home/ferperez/SigMA_tests/Input_datasets/BRCA/MSK_IMPACTv1', '/home/ferperez/SigMA_tests/Input_datasets/BRCA/MSK_IMPACTv2', '/home/ferperez/SigMA_tests/Input_datasets/PAAD/mc3', '/home/ferperez/SigMA_tests/Input_datasets/PAAD/SeqCap_capture_targets', '/home/ferperez/SigMA_tests/Input_datasets/COADREAD/Foundation_One', '/home/ferperez/SigMA_tests/Input_datasets/COADREAD/MSK_IMPACTv1', '/home/ferperez/SigMA_tests/Input_datasets/COADREAD/MSK_IMPACTv2']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import time\n",
    "import warnings\n",
    "import sigmapy as sgpy\n",
    "import numpy as np\n",
    "import glob\n",
    "\n",
    "# Project paths\n",
    "project_folder = \"/home/ferperez/SigMA_tests/\"\n",
    "input_path = \"/home/ferperez/SigMA_tests/Input_datasets/\"\n",
    "output_path = \"/home/ferperez/SigMA_tests/Output_SigMA_Py/\"\n",
    "\n",
    "# Listing subdirectories in Input_datasets\n",
    "all_dirs = [os.path.join(dp, d) for dp, dn, filenames in os.walk(input_path) for d in dn]\n",
    "len_dir_depth = [len(path.split(\"/\")) for path in all_dirs]\n",
    "max_depth = max(len_dir_depth)\n",
    "input_dirs = [d for d, depth in zip(all_dirs, len_dir_depth) if depth == max_depth]\n",
    "\n",
    "print(\"Input directories:\")\n",
    "print(input_dirs)\n",
    "\n",
    "#Creating dictionary for matching input data with SigMA datasets\n",
    "samples_dict = pd.DataFrame({\n",
    "    \"tumor_folder\": [\"BRCA\", \"COADREAD\", \"PAAD\"],\n",
    "    \"tumor_abreviation\": [\"breast\", \"crc\", \"panc_ad\"]\n",
    "})\n",
    "\n",
    "sequencing_dict = pd.DataFrame({\n",
    "    \"seq_folder\": [\"MSK_IMPACTv1\", \"mc3\"],\n",
    "    \"seq_abreviation\": [\"msk\", \"tcga_mc3\"]\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0f81f61b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting to analyse: /home/ferperez/SigMA_tests/Input_datasets/BRCA/mc3\n",
      "96-dimensional matrix is saved in /home/ferperez/SigMA_tests/Output_SigMA_Py/BRCA/mc3/Mut_counts.csv\n",
      "Running SigMA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "package ‘gbm’ was built under R version 4.3.3 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SigMA analysis completed in 107.47 seconds.\n",
      "Starting to analyse: /home/ferperez/SigMA_tests/Input_datasets/BRCA/MSK_IMPACTv1\n",
      "96-dimensional matrix is saved in /home/ferperez/SigMA_tests/Output_SigMA_Py/BRCA/MSK_IMPACTv1/Mut_counts.csv\n",
      "Running SigMA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "package ‘gbm’ was built under R version 4.3.3 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SigMA analysis completed in 22.09 seconds.\n",
      "Starting to analyse: /home/ferperez/SigMA_tests/Input_datasets/PAAD/mc3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ferperez/SigMApy/sigmapy/matrix.py:606: DtypeWarning: Columns (4,38,57,63,74,78) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  maf = pd.read_csv(files[0], comment='#', sep='\\t', header=0, encoding='utf-8')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "96-dimensional matrix is saved in /home/ferperez/SigMA_tests/Output_SigMA_Py/PAAD/mc3/Mut_counts.csv\n",
      "Running SigMA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ferperez/SigMApy/sigmapy/matrix.py:606: DtypeWarning: Columns (4,38,57,63,74,78) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  maf = pd.read_csv(files[0], comment='#', sep='\\t', header=0, encoding='utf-8')\n",
      "Warning message:\n",
      "package ‘gbm’ was built under R version 4.3.3 \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SigMA analysis completed in 18.80 seconds.\n",
      "Starting to analyse: /home/ferperez/SigMA_tests/Input_datasets/COADREAD/MSK_IMPACTv1\n",
      "96-dimensional matrix is saved in /home/ferperez/SigMA_tests/Output_SigMA_Py/COADREAD/MSK_IMPACTv1/Mut_counts.csv\n",
      "Running SigMA\n",
      "SigMA analysis completed in 13.57 seconds.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error in eval(predvars, data, env) : \n",
      "  object 'Signature_3_c1_ml_ml' not found\n",
      "Calls: source ... predict.gbm -> model.frame -> model.frame.default -> eval -> eval\n",
      "In addition: Warning message:\n",
      "package ‘gbm’ was built under R version 4.3.3 \n",
      "Execution halted\n",
      "/home/ferperez/SigMApy/sigmapy/tools/mva_predictor.py:58: UserWarning: MVA prediction failed: Command '['Rscript', '-e', '\\n            args <- c(\"/tmp/tmpw3kkwl51.csv\", \"/tmp/tmpqmq5m1gz.csv\",\"Signature_3\", \"msk\", \"crc\", \"0\", \"0\", \"/home/ferperez/SigMApy/sigmapy/var/sysdata.rda\");\\n            \\n            source(\"/home/ferperez/SigMApy/sigmapy/r/predict_mva.R\");\\n            ']' returned non-zero exit status 1.\n",
      "  warnings.warn(f\"MVA prediction failed: {e}\")\n"
     ]
    }
   ],
   "source": [
    "# Running SigMA directory by directory\n",
    "for mydir in input_dirs:\n",
    "    parts = mydir.split(\"/\")\n",
    "    seq_type = parts[-1]\n",
    "    sample_type = parts[-2]\n",
    "\n",
    "    seq_type_for_SigMA = sequencing_dict.loc[sequencing_dict[\"seq_folder\"] == seq_type, \"seq_abreviation\"]\n",
    "    sample_type_for_SigMA = samples_dict.loc[samples_dict[\"tumor_folder\"] == sample_type, \"tumor_abreviation\"]\n",
    "\n",
    "    # In case the input folders are not supported by SigMA\n",
    "    if seq_type_for_SigMA.empty or sample_type_for_SigMA.empty:\n",
    "        #print(\"Ignoring the next dir because of lack of matching in list_tumor_types()\")\n",
    "        #print(mydir)\n",
    "        continue\n",
    "    print(f\"Starting to analyse: {mydir}\")\n",
    "\n",
    "    # Defining the output folder for storing results\n",
    "    output_dir_iter = os.path.join(output_path, sample_type, seq_type)\n",
    "    os.makedirs(output_dir_iter, exist_ok=True)\n",
    "\n",
    "    seq_type_for_SigMA = seq_type_for_SigMA.values[0]\n",
    "    sample_type_for_SigMA = sample_type_for_SigMA.values[0]\n",
    "\n",
    "    # The datasets in mc3 folder are MAF, otherwise VCF\n",
    "    if seq_type_for_SigMA == \"tcga_mc3\":\n",
    "        data_type = \"maf\"\n",
    "        mydir = os.path.join(mydir, f\"{sample_type_for_SigMA}.maf\")\n",
    "    else:\n",
    "        data_type = \"vcf\"\n",
    "\n",
    "    # Generating count matrix\n",
    "    #Running matrix class from sgpy\n",
    "    matrix_instance = sgpy.Matrix(mydir,\n",
    "                                    ref_genome=\"hg19\")\n",
    "    mut_counts_file = os.path.join(output_dir_iter, \"Mut_counts.csv\")\n",
    "\n",
    "    matrix_instance.df.to_csv(mut_counts_file, sep=\",\", index=True)\n",
    "    print(f\"96-dimensional matrix is saved in {mut_counts_file}\")\n",
    "    \n",
    "    print(\"Running SigMA\")\n",
    "\n",
    "    start_time = time.perf_counter()\n",
    "\n",
    "    sigMA_output = sgpy.run(\n",
    "                    input=mydir,\n",
    "                    tumor_type=sample_type_for_SigMA,\n",
    "                    ref_genome=\"hg19\",\n",
    "                    mxw=12,\n",
    "                    soi=\"SBS3\",\n",
    "                    mva=True,\n",
    "                    seq_type=seq_type_for_SigMA,\n",
    "                    catalog=\"cosmic_v2_inhouse\",\n",
    "                    msi=False)\n",
    "    \n",
    "    sigMA_output.df.to_csv(os.path.join(output_dir_iter, \"SigMA_output.csv\"), sep=\",\", index=True)\n",
    "    end_time = time.perf_counter()\n",
    "    print(f\"SigMA analysis completed in {end_time - start_time:.2f} seconds.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab031166",
   "metadata": {},
   "source": [
    "### Comparing now the results from SigMApy vs SigMAR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a54b84a",
   "metadata": {},
   "source": [
    "Auxiliar functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "50113c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to compare the differences between two SNV matrixs when they are not equal\n",
    "def compare_dataframes(df1, df2):\n",
    "    # Ensure same index and columns order\n",
    "    df1 = df1.sort_index().sort_index(axis=1)\n",
    "    df2 = df2.sort_index().sort_index(axis=1)\n",
    "\n",
    "    # Check shape and columns\n",
    "    if not (df1.shape == df2.shape and (df1.columns == df2.columns).all() and (df1.index == df2.index).all()):\n",
    "        print(\"DataFrames do not have the same shape, columns, or index.\")\n",
    "        return\n",
    "    \n",
    "    # Find differences\n",
    "    diff = df1 != df2\n",
    "    n_diff = diff.values.sum()\n",
    "    print(f\"Number of discordant values: {n_diff}\")\n",
    "    if n_diff > 0:\n",
    "        rows, cols = diff.to_numpy().nonzero()\n",
    "        for r, c in zip(rows, cols):\n",
    "            print(f\"Row: {df1.index[r]}, Column: {df1.columns[c]}, SigmaPY: {df1.iat[r, c]}, SigmaR: {df2.iat[r, c]}\")\n",
    "    else:\n",
    "        print(\"All values are concordant.\")\n",
    "\n",
    "\n",
    "\n",
    "# Comparing the outputs from SigMA Python and SigMA R\n",
    "def mean_diff_exps(df1, df2):\n",
    "    # Calculate mean absolute difference for 'exps_all' columns\n",
    "    diffs = []\n",
    "    for py, r in zip(df1['exps_all'], df2['exps_all']):\n",
    "        py_vals = np.array([float(x) for x in str(py).split('_')])\n",
    "        r_vals = np.array([float(x) for x in str(r).split('_')])\n",
    "        # Ensure same length\n",
    "        if len(py_vals) == len(r_vals):\n",
    "            diffs.extend(np.abs(py_vals - r_vals))\n",
    "        else:\n",
    "            print(\"Row length mismatch, skipping.\")\n",
    "    mean_diff = np.mean(diffs)\n",
    "    return(mean_diff)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14aac4f",
   "metadata": {},
   "source": [
    "##### Comparing if the SNV count matrix are the same"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "77990382",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ferperez/SigMA_tests/Input_datasets/BRCA/mc3\n",
      "Are the 96-dimensional matrices from SigMA_Python and SigMA_R equal? True\n",
      "/home/ferperez/SigMA_tests/Input_datasets/BRCA/MSK_IMPACTv1\n",
      "Are the 96-dimensional matrices from SigMA_Python and SigMA_R equal? True\n",
      "/home/ferperez/SigMA_tests/Input_datasets/PAAD/mc3\n",
      "Are the 96-dimensional matrices from SigMA_Python and SigMA_R equal? False\n",
      "Comparing the discordant values:\n",
      "Number of discordant values: 91\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: caaa, SigmaPY: 0, SigmaR: 3\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: caac, SigmaPY: 11, SigmaR: 71\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: caag, SigmaPY: 2, SigmaR: 4\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: caat, SigmaPY: 51, SigmaR: 246\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: caca, SigmaPY: 12, SigmaR: 65\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cacc, SigmaPY: 26, SigmaR: 161\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cacg, SigmaPY: 3, SigmaR: 32\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cact, SigmaPY: 206, SigmaR: 1086\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: caga, SigmaPY: 6, SigmaR: 30\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cagc, SigmaPY: 34, SigmaR: 208\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cagg, SigmaPY: 9, SigmaR: 30\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cagt, SigmaPY: 215, SigmaR: 1151\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cata, SigmaPY: 5, SigmaR: 31\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: catc, SigmaPY: 41, SigmaR: 246\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: catg, SigmaPY: 1, SigmaR: 7\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: catt, SigmaPY: 171, SigmaR: 963\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgaa, SigmaPY: 1, SigmaR: 7\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgac, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgag, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgat, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgca, SigmaPY: 0, SigmaR: 1\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgcc, SigmaPY: 0, SigmaR: 1\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgct, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgga, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cggt, SigmaPY: 4, SigmaR: 6\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgta, SigmaPY: 1, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgtc, SigmaPY: 0, SigmaR: 1\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cgtt, SigmaPY: 2, SigmaR: 3\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctaa, SigmaPY: 45, SigmaR: 239\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctac, SigmaPY: 63, SigmaR: 373\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctag, SigmaPY: 205, SigmaR: 1238\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctat, SigmaPY: 70, SigmaR: 377\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctca, SigmaPY: 19, SigmaR: 116\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctcc, SigmaPY: 70, SigmaR: 305\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctcg, SigmaPY: 181, SigmaR: 1102\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctct, SigmaPY: 67, SigmaR: 397\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctga, SigmaPY: 108, SigmaR: 572\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctgc, SigmaPY: 214, SigmaR: 1080\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctgg, SigmaPY: 522, SigmaR: 2825\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctgt, SigmaPY: 211, SigmaR: 1097\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: ctta, SigmaPY: 30, SigmaR: 159\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cttc, SigmaPY: 95, SigmaR: 534\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cttg, SigmaPY: 238, SigmaR: 1232\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: cttt, SigmaPY: 108, SigmaR: 590\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: taaa, SigmaPY: 1, SigmaR: 8\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: taac, SigmaPY: 4, SigmaR: 24\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: taag, SigmaPY: 1, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: taat, SigmaPY: 4, SigmaR: 33\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: taca, SigmaPY: 0, SigmaR: 1\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tacc, SigmaPY: 6, SigmaR: 29\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tacg, SigmaPY: 0, SigmaR: 1\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tact, SigmaPY: 4, SigmaR: 17\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: taga, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tagc, SigmaPY: 7, SigmaR: 30\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tagt, SigmaPY: 3, SigmaR: 14\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tata, SigmaPY: 0, SigmaR: 2\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tatc, SigmaPY: 0, SigmaR: 14\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tatg, SigmaPY: 1, SigmaR: 5\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tatt, SigmaPY: 1, SigmaR: 15\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcaa, SigmaPY: 21, SigmaR: 108\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcac, SigmaPY: 24, SigmaR: 126\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcag, SigmaPY: 21, SigmaR: 129\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcat, SigmaPY: 17, SigmaR: 94\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcca, SigmaPY: 5, SigmaR: 30\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tccc, SigmaPY: 11, SigmaR: 80\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tccg, SigmaPY: 21, SigmaR: 126\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcct, SigmaPY: 6, SigmaR: 66\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcga, SigmaPY: 46, SigmaR: 253\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcgc, SigmaPY: 56, SigmaR: 306\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcgg, SigmaPY: 40, SigmaR: 239\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcgt, SigmaPY: 38, SigmaR: 207\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tcta, SigmaPY: 4, SigmaR: 50\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tctc, SigmaPY: 18, SigmaR: 67\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tctg, SigmaPY: 19, SigmaR: 83\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tctt, SigmaPY: 23, SigmaR: 112\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgaa, SigmaPY: 1, SigmaR: 4\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgac, SigmaPY: 13, SigmaR: 65\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgag, SigmaPY: 4, SigmaR: 23\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgat, SigmaPY: 36, SigmaR: 149\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgca, SigmaPY: 1, SigmaR: 7\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgcc, SigmaPY: 12, SigmaR: 55\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgcg, SigmaPY: 8, SigmaR: 47\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgct, SigmaPY: 39, SigmaR: 200\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgga, SigmaPY: 5, SigmaR: 13\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tggc, SigmaPY: 10, SigmaR: 55\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tggg, SigmaPY: 7, SigmaR: 41\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tggt, SigmaPY: 27, SigmaR: 123\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgta, SigmaPY: 0, SigmaR: 7\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgtc, SigmaPY: 10, SigmaR: 89\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgtg, SigmaPY: 6, SigmaR: 18\n",
      "Row: TCGA-IB-7651-01A-11D-2154-08, Column: tgtt, SigmaPY: 43, SigmaR: 285\n",
      "/home/ferperez/SigMA_tests/Input_datasets/COADREAD/MSK_IMPACTv1\n",
      "Are the 96-dimensional matrices from SigMA_Python and SigMA_R equal? True\n"
     ]
    }
   ],
   "source": [
    "#Reading the output files from SigMA R version and Python verision\n",
    "\n",
    "R_output_path = \"/home/ferperez/SigMA_tests/Output_SigMA_Rversion/\"\n",
    "\n",
    "\n",
    "for mydir in input_dirs:\n",
    "    parts = mydir.split(\"/\")\n",
    "    seq_type = parts[-1]\n",
    "    sample_type = parts[-2]\n",
    "\n",
    "    seq_type_for_SigMA = sequencing_dict.loc[sequencing_dict[\"seq_folder\"] == seq_type, \"seq_abreviation\"]\n",
    "    sample_type_for_SigMA = samples_dict.loc[samples_dict[\"tumor_folder\"] == sample_type, \"tumor_abreviation\"]\n",
    "\n",
    "    # In case the input folders are not supported by SigMA\n",
    "    if seq_type_for_SigMA.empty or sample_type_for_SigMA.empty:\n",
    "        continue\n",
    "    \n",
    "    print(mydir)\n",
    "\n",
    "    # Reading and sorting the SNV count matrix generated by SigMA Python version\n",
    "    output_dir_iter = os.path.join(output_path, sample_type, seq_type)\n",
    "    SigmaPY_matrix_out = pd.read_csv(os.path.join(output_dir_iter, \"Mut_counts.csv\"))\n",
    "    SigmaPY_matrix_out.index = SigmaPY_matrix_out['Unnamed: 0'].apply(os.path.basename)\n",
    "    SigmaPY_matrix_out = SigmaPY_matrix_out.drop('Unnamed: 0', axis=1)\n",
    "    SigmaPY_matrix_out = SigmaPY_matrix_out.sort_index().sort_index(axis=1)\n",
    "\n",
    "    # Reading and sorting the SNV count matrix generated by SigMA R version\n",
    "    R_output_dir_iter = os.path.join(R_output_path, sample_type, seq_type)\n",
    "    SigmaR_matrix_out = pd.read_csv(os.path.join(R_output_dir_iter, \"Mut_counts.csv\"))\n",
    "    SigmaR_matrix_out.index = SigmaR_matrix_out['tumor'].apply(os.path.basename)\n",
    "    SigmaR_matrix_out = SigmaR_matrix_out.drop('tumor', axis=1)\n",
    "    SigmaR_matrix_out = SigmaR_matrix_out.sort_index().sort_index(axis=1)\n",
    "\n",
    "    Are_equal = SigmaPY_matrix_out.equals(SigmaR_matrix_out)\n",
    "    \n",
    "    print(f\"Are the 96-dimensional matrices from SigMA_Python and SigMA_R equal? {Are_equal}\")\n",
    "\n",
    "    if (Are_equal == False):\n",
    "        print(\"Comparing the discordant values:\")\n",
    "        compare_dataframes(SigmaPY_matrix_out, SigmaR_matrix_out)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da161da3",
   "metadata": {},
   "source": [
    "##### Comparing how similar the exposures and Sig3 clasifications are"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "61ed6874",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ferperez/SigMA_tests/Input_datasets/BRCA/mc3\n",
      "/home/ferperez/SigMA_tests/Input_datasets/BRCA/MSK_IMPACTv1\n",
      "/home/ferperez/SigMA_tests/Input_datasets/PAAD/mc3\n",
      "/home/ferperez/SigMA_tests/Input_datasets/COADREAD/MSK_IMPACTv1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Input_folder</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>N_samples_R</th>\n",
       "      <th>N_samples_Python</th>\n",
       "      <th>Concordant_sigs</th>\n",
       "      <th>Disconcordant_sigs</th>\n",
       "      <th>Mean_exp_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/home/ferperez/SigMA_tests/Input_datasets/BRCA...</td>\n",
       "      <td>tcga_mc3</td>\n",
       "      <td>breast</td>\n",
       "      <td>985</td>\n",
       "      <td>982</td>\n",
       "      <td>982</td>\n",
       "      <td>0</td>\n",
       "      <td>5.358043e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/home/ferperez/SigMA_tests/Input_datasets/BRCA...</td>\n",
       "      <td>msk</td>\n",
       "      <td>breast</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>127</td>\n",
       "      <td>0</td>\n",
       "      <td>1.176575e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/home/ferperez/SigMA_tests/Input_datasets/PAAD...</td>\n",
       "      <td>tcga_mc3</td>\n",
       "      <td>panc_ad</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>149</td>\n",
       "      <td>0</td>\n",
       "      <td>2.074130e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/home/ferperez/SigMA_tests/Input_datasets/COAD...</td>\n",
       "      <td>msk</td>\n",
       "      <td>crc</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>139</td>\n",
       "      <td>0</td>\n",
       "      <td>1.251579e-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Input_folder  seq_type sample_type  \\\n",
       "0  /home/ferperez/SigMA_tests/Input_datasets/BRCA...  tcga_mc3      breast   \n",
       "1  /home/ferperez/SigMA_tests/Input_datasets/BRCA...       msk      breast   \n",
       "2  /home/ferperez/SigMA_tests/Input_datasets/PAAD...  tcga_mc3     panc_ad   \n",
       "3  /home/ferperez/SigMA_tests/Input_datasets/COAD...       msk         crc   \n",
       "\n",
       "   N_samples_R  N_samples_Python  Concordant_sigs  Disconcordant_sigs  \\\n",
       "0          985               982              982                   0   \n",
       "1          127               127              127                   0   \n",
       "2          149               149              149                   0   \n",
       "3          139               139              139                   0   \n",
       "\n",
       "   Mean_exp_difference  \n",
       "0         5.358043e-14  \n",
       "1         1.176575e-14  \n",
       "2         2.074130e+01  \n",
       "3         1.251579e-14  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Reading the output files from SigMA R version and Python verision\n",
    "\n",
    "R_output_path = \"/home/ferperez/SigMA_tests/Output_SigMA_Rversion/\"\n",
    "comparison_results = pd.DataFrame(columns=[\"Input_folder\", \"seq_type\",\n",
    "                                            \"sample_type\", \"N_samples_R\",\n",
    "                                            \"N_samples_Python\",\n",
    "                                            \"Concordant_sigs\", \"Disconcordant_sigs\",\n",
    "                                            \"Mean_exp_difference\"])\n",
    "\n",
    "\n",
    "for mydir in input_dirs:\n",
    "    parts = mydir.split(\"/\")\n",
    "    seq_type = parts[-1]\n",
    "    sample_type = parts[-2]\n",
    "\n",
    "    seq_type_for_SigMA = sequencing_dict.loc[sequencing_dict[\"seq_folder\"] == seq_type, \"seq_abreviation\"]\n",
    "    sample_type_for_SigMA = samples_dict.loc[samples_dict[\"tumor_folder\"] == sample_type, \"tumor_abreviation\"]\n",
    "\n",
    "    # In case the input folders are not supported by SigMA\n",
    "    if seq_type_for_SigMA.empty or sample_type_for_SigMA.empty:\n",
    "        continue\n",
    "    \n",
    "    print(mydir)\n",
    "\n",
    "    # Reading and sorting the SNV count matrix generated by SigMA Python version\n",
    "    output_dir_iter = os.path.join(output_path, sample_type, seq_type)\n",
    "    SigmaPY_out = pd.read_csv(os.path.join(output_dir_iter, \"SigMA_output.csv\"))\n",
    "    SigmaPY_out.index = SigmaPY_out['Unnamed: 0'].apply(os.path.basename)\n",
    "    SigmaPY_out = SigmaPY_out.drop('Unnamed: 0', axis=1)\n",
    "    SigmaPY_out = SigmaPY_out.sort_index().sort_index(axis=1)\n",
    "\n",
    "    # Reading and sorting the SNV count matrix generated by SigMA R version\n",
    "    R_output_dir_iter = os.path.join(R_output_path, sample_type, seq_type)\n",
    "    lite_files = glob.glob(os.path.join(R_output_dir_iter, \"*_lite.csv\"))\n",
    "    SigmaR_out = pd.read_csv(lite_files[0])\n",
    "    SigmaR_out.index = SigmaR_out['tumor'].apply(os.path.basename)\n",
    "    SigmaR_out = SigmaR_out.drop('tumor', axis=1)\n",
    "    SigmaR_out = SigmaR_out.sort_index().sort_index(axis=1)\n",
    "\n",
    "    #Countin how many samples are per result,\n",
    "    #The results does not always have the same samples\n",
    "    N_samples_R = SigmaR_out.shape[0]\n",
    "    N_samples_Python = SigmaPY_out.shape[0]\n",
    "    common_idx = SigmaPY_out.index.intersection(SigmaR_out.index)\n",
    "\n",
    "\n",
    "    Concordant_samples = (SigmaPY_out.loc[common_idx, 'sigs_all'] == SigmaR_out.loc[common_idx, 'sigs_all']).sum()\n",
    "    Disconcordant_samples = (SigmaPY_out.loc[common_idx, 'sigs_all'] != SigmaR_out.loc[common_idx, 'sigs_all']).sum()\n",
    "\n",
    "\n",
    "    mean_difference = mean_diff_exps(SigmaPY_out.loc[common_idx], SigmaR_out.loc[common_idx])\n",
    "\n",
    "\n",
    "    # Store results in a DataFrame\n",
    "    comparison_results.loc[len(comparison_results)] = [mydir, seq_type_for_SigMA.values[0],\n",
    "                                                        sample_type_for_SigMA.values[0],\n",
    "                                                        N_samples_R, N_samples_Python,\n",
    "                                                        Concordant_samples,\n",
    "                                                        Disconcordant_samples,\n",
    "                                                        mean_difference]\n",
    "comparison_results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sigmapy310",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
